File: app/src/main/java/net/sarasarasa/basemvp/base/MvpFragment.java
package net.sarasarasa.basemvp.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

@SuppressWarnings("unchecked")
public abstract class MvpFragment<P extends MvpPresenter> extends Fragment implements IMvpView {

    protected P mPresenter;
    protected boolean isRetainInstance = false;
    private View rootView;

    @LayoutRes
    abstract int getLayoutResId();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPresenter = createPresenter();
        if (mPresenter == null) {
            throw new NullPointerException("Presenter is null");
        }

        initView();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle state) {
        rootView = inflater.inflate(getLayoutResId(), container, false);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (mPresenter != null) {
            mPresenter.attachView(this);
        }

        initView();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (mPresenter != null) {
            mPresenter.detachView(isRetainInstance);
        }

    }

    /**
     * Instantiate a mPresenter instance
     *
     * @return The {@link MvpPresenter} for this view
     */
    protected abstract P createPresenter();

    protected abstract void initView();
}

File: kotlin/src/main/java/net/sarasarasa/basemvpkotlin/base/MvpFragment.kt
package net.sarasarasa.basemvpkotlin.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.annotation.LayoutRes
import androidx.fragment.app.Fragment

@Suppress("UNCHECKED_CAST")
abstract class MvpFragment<in V : IMvpView, P : IMvpPresenter<V>> : Fragment(), IMvpView {

    protected var mPresenter: P? = null
    protected val isRetainInstance = false

    private var rootView: View? = null

    @get:LayoutRes
    internal abstract val layoutResId: Int

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mPresenter = createPresenter()
        if (mPresenter == null) {
            throw NullPointerException("Presenter is null")
        }

        initView()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, state: Bundle?): View? {
        rootView = inflater.inflate(layoutResId, container, false)
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mPresenter?.attachView(this as V)
        initView()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter?.detachView(isRetainInstance)
    }

    /**
     * Instantiate a mPresenter instance
     *
     * @return The [MvpPresenter] for this view
     */
    protected abstract fun createPresenter(): P

    protected abstract fun initView()
}